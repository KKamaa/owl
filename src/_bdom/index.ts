import { Block } from "./types";

export { makeBlock } from "./element";
export { list } from "./list";
export { multi } from "./multi";
export { text } from "./text";

// export function beforeRemove(block: Block) {}

// -----------------------------------------------------------------------------
//  BDom main entry points
// -----------------------------------------------------------------------------

export function mount(block: Block, target: HTMLElement) {
  const anchor = document.createTextNode("");
  target.appendChild(anchor);
  block.ops.mountBefore(block, anchor);
  anchor.remove();
}

export function patch(block1: Block, block2: Block) {
  block1.ops.patch(block1, block2);
}

export function remove(block: Block) {
  // beforeRemove(block);
  block.ops.remove(block);
}
