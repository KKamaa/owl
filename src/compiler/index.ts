import { setupBlockDom } from "./helpers";
import { QWebCompiler, TemplateFunction } from "./qweb_compiler";
export { Template, TemplateFunction } from "./qweb_compiler";

setupBlockDom();

export function compileTemplate(template: string, name?: string): TemplateFunction {
  const compiler = new QWebCompiler(template, name);
  return compiler.compile();
}
