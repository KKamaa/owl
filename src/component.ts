import type { BNode } from "./b_node";

export class Component {
  static template: string = "";

  props: any;
  env: any;
  __owl__: BNode;

  constructor(props: any, env: any, node: BNode) {
    this.props = props;
    this.env = env;
    this.__owl__ = node;
  }
  get el(): HTMLElement | null {
    const node = this.__owl__;
    return node.bdom ? (node.bdom.el as any) : null;
  }

  setup() {}

  render(): Promise<void> {
    return this.__owl__.render();
  }
}
