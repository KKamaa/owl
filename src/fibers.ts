import type { Block } from "./bdom";
import { STATUS } from "./status";
import { OwlNode } from "./owl_node";

export function makeChildFiber(node: OwlNode, parent: Fiber): Fiber {
  // todo: remove this
  if ((window as any).debug) {
    console.log("childfiber for " + node.component.constructor.name);
  }
  let current = node.fiber;
  if (current) {
    // current is necessarily a rootfiber here
    let root = parent.root;
    cancelFibers(current.children);
    current.children = [];
    current.parent = parent;
    root.counter++;
    return current;
  }
  let result = new Fiber(node, parent);
  return result;
}

export function makeRootFiber(node: OwlNode): Fiber {
  // todo: remove this
  if ((window as any).debug) {
    console.log("rootfiber for " + node.component.constructor.name);
  }
  let current = node.fiber;
  if (current) {
    let root = current.root;
    root.counter -= cancelFibers(current.children);
    current.children = [];
    root.counter++;
    current.bdom = null;
    return current;
  }
  let result = new RootFiber(node);
  return result;
}

export function makeMountFiber(node: OwlNode, target: HTMLElement): MountFiber {
  if ((window as any).debug) {
    console.log("mountfiber for " + node.component.constructor.name);
  }
  let result = new MountFiber(node, target);
  return result;
}

/**
 * @returns number of not-yet rendered fibers cancelled
 */
function cancelFibers(fibers: Fiber[]): number {
  let result = 0;
  for (let fiber of fibers) {
    fiber.node.fiber = null;
    if (!fiber.bdom) {
      result++;
    }
    result += cancelFibers(fiber.children);
  }
  return result;
}

export class Fiber {
  node: OwlNode;
  bdom: Block | null = null;
  root: RootFiber;
  parent: Fiber | null;
  children: Fiber[] = [];

  constructor(node: OwlNode, parent: Fiber | null) {
    this.node = node;
    node.fiber = this;
    this.parent = parent;
    if (parent) {
      const root = parent.root;
      root.counter++;
      this.root = root;
      parent.children.push(this);
    } else {
      this.root = this as any;
    }
  }
}

export class RootFiber extends Fiber {
  counter: number = 1;
  error: Error | null = null;
  resolve: any;
  promise: any;
  reject: any;
  toPatch: Fiber[];

  constructor(node: OwlNode) {
    super(node, null);
    this.counter = 1;
    this.error = null;
    this.toPatch = [this];

    this.promise = new Promise((resolve, reject) => {
      this.resolve = resolve;
      this.reject = reject;
    });
  }

  complete() {
    const node = this.node;
    // if (node.fiber !== fiber) {
    //   return;
    // }
    for (let fiber of this.toPatch) {
      let node = fiber.node;
      // because of the asynchronous nature of the rendering, some parts of the
      // UI may have been rendered, then deleted in a followup rendering, and we
      // do not want to call onBeforePatch in that case.
      if (node.fiber === fiber) {
        node.callBeforePatch();
      }
    }
    const mountedNodes: any[] = [];
    const patchedNodes: any[] = [node];
    node.bdom!.patch(this.bdom!, mountedNodes, patchedNodes);
    this.finalize(mountedNodes, patchedNodes);
    node.fiber = null;
  }

  finalize(mounted: OwlNode[], patched: OwlNode[]) {
    let current;
    while ((current = mounted.pop())) {
      for (let cb of current.mounted) {
        cb();
      }
    }
    while ((current = patched.pop())) {
      for (let cb of current.patched) {
        cb();
      }
    }
  }
}

export class MountFiber extends RootFiber {
  target: HTMLElement;

  constructor(node: OwlNode, target: HTMLElement) {
    super(node);
    this.target = target;
    this.toPatch = [];
  }
  complete() {
    const node = this.node;
    node.bdom = this.bdom;
    const mountedNodes: any[] = [node];
    const patchedNodes: any[] = [];
    node.bdom!.mount(this.target, mountedNodes, patchedNodes);
    this.finalize(mountedNodes, patchedNodes);
    node.status = STATUS.MOUNTED;
    node.fiber = null;
  }
}
