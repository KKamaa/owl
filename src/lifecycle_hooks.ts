import { getCurrent } from "./core/owl_node";

// -----------------------------------------------------------------------------
//  hooks
// -----------------------------------------------------------------------------

export function onWillStart(fn: () => Promise<void> | void) {
  const node = getCurrent()!;
  node.willStart.push(fn);
}

export function onBeforeUnmount(fn: () => Promise<void> | void) {
  const node = getCurrent()!;
  node.beforeUnmount.push(fn);
}

export function onMounted(fn: () => void) {
  const node = getCurrent()!;
  node.mounted.push(fn);
}

export function onBeforePatch(fn: () => Promise<void> | void) {
  const node = getCurrent()!;
  node.beforePatch.push(fn);
}

export function onPatched(fn: () => void) {
  const node = getCurrent()!;
  node.patched.push(fn);
}
