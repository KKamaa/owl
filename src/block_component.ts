import { Block } from "./bdom";
import { makeChildFiber, __internal__destroyed } from "./fibers";
import { OwlNode } from "./owl_node";
import { STATUS } from "./status";

// -----------------------------------------------------------------------------
//  Component Block
// -----------------------------------------------------------------------------

export class BComponent extends Block {
  node: OwlNode;
  parentClass?: any;
  classTarget?: HTMLElement;

  /**
   *
   * @param name
   * @param props
   * @param key
   * @param owner the component in which the component was defined
   * @param parent the actual parent (may be different in case of slots)
   */
  constructor(name: string, props: any, key: string, owner: any, parent: any) {
    super();
    const parentNode: OwlNode = parent.__owl__;
    let node: OwlNode | undefined = parentNode.children[key];
    if (node && node.status < STATUS.MOUNTED) {
      node.destroy();
      // delete parentNode.children[key];
      node = undefined;
    }
    const parentFiber = parentNode.fiber!;
    if (node) {
      // update
      const fiber = makeChildFiber(node, parentFiber);
      if (node.willPatch.length) {
        parentFiber.root.willPatch.push(fiber);
      }
      if (node.patched.length) {
        parentFiber.root.patched.push(fiber);
      }
      node.updateAndRender(props, fiber);
    } else {
      // new component
      const components = owner.constructor.components;
      const C = components[name];
      node = new OwlNode(parentNode.app, C, props);
      parentNode.children[key] = node;
      const fiber = makeChildFiber(node, parentNode.fiber!);
      if (node.mounted.length) {
        parentFiber.root.mounted.push(fiber);
      }
      node.initiateRender(fiber);
    }
    this.node = node;
  }

  firstChildNode(): ChildNode | null {
    const bdom = this.node.bdom;
    return bdom ? bdom.firstChildNode() : null;
  }

  mountBefore(anchor: ChildNode) {
    const node = this.node;
    const bdom = node.fiber!.bdom!;
    node.bdom = bdom;
    bdom.mountBefore(anchor);
    if (this.parentClass) {
      this.parentClass = this.parentClass.trim().split(/\s+/);
      const el = this.firstChildNode();
      if (el instanceof HTMLElement) {
        this.addClass(el);
      }
    }
    node.status = STATUS.MOUNTED;
    node.fiber!.appliedToDom = true;
    node.fiber = null;
  }

  moveBefore(anchor: ChildNode) {
    this.node.bdom!.moveBefore(anchor);
  }

  addClass(el: HTMLElement) {
    this.classTarget = el;
    for (let cl of this.parentClass) {
      el.classList.add(cl);
    }
  }

  removeClass(el: HTMLElement) {
    for (let cl of this.parentClass) {
      el.classList.remove(cl);
    }
  }

  patch() {
    const node = this.node;

    node.bdom!.patch(node!.fiber!.bdom!);
    if (this.parentClass) {
      const el = this.firstChildNode();
      if (el !== this.classTarget) {
        if (el && this.classTarget) {
          this.removeClass(this.classTarget);
          this.addClass(el as any);
        } else if (el) {
          this.addClass(el as any);
        } else {
          this.removeClass(this.classTarget!);
          this.classTarget = undefined;
        }
      }
    }
    node.fiber!.appliedToDom = true;
    node.fiber = null;
  }

  beforeRemove() {
    visitRemovedNodes(this.node);
  }

  remove() {
    const bdom = this.node.bdom!;
    bdom.remove();
  }
}

function visitRemovedNodes(node: OwlNode) {
  if (node.status === STATUS.MOUNTED) {
    const component = node.component;
    for (let cb of node.willUnmount) {
      cb.call(component);
    }
  }
  for (let child of Object.values(node.children)) {
    visitRemovedNodes(child);
  }
  node.status = STATUS.DESTROYED;
  if (node.destroyed.length) {
    __internal__destroyed.push(node);
  }
}

export class BComponentH extends BComponent {
  handlers: any[];
  constructor(handlers: number, name: string, props: any, key: string, owner: any, parent: any) {
    super(name, props, key, owner, parent);
    this.handlers = new Array(handlers);
  }
  mountBefore(anchor: ChildNode) {
    super.mountBefore(anchor);
    this.setupHandlers();
  }
  setupHandlers() {
    for (let i = 0; i < this.handlers.length; i++) {
      const handler = this.handlers[i];
      const eventType = handler[0];
      const el = this.node.component.el!;
      el.addEventListener(eventType, (ev: Event) => {
        const info = this.handlers![i];
        const [, callback] = info;
        callback(ev);
      });
    }
  }
}
