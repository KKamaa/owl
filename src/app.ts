import { Blocks as BaseBlocks } from "./bdom";
import { compileTemplate, Template } from "./compiler/index";
import { globalTemplates } from "./tags";
import { BComponent, BComponentH } from "./block_component";
import type { Component } from "./component";
import { OwlNode } from "./owl_node";
import { Scheduler } from "./scheduler";
import { UTILS } from "./template_utils";

const Blocks = {
  ...BaseBlocks,
  BComponent,
  BComponentH,
};

// -----------------------------------------------------------------------------
//  TemplateSet
// -----------------------------------------------------------------------------

export class TemplateSet {
  rawTemplates: { [name: string]: string } = Object.create(globalTemplates);
  templates: { [name: string]: Template } = {};
  utils: typeof UTILS;

  constructor() {
    const call = (subTemplate: string, ctx: any, parent: any) => {
      const template = this.getTemplate(subTemplate);
      return template(ctx, parent);
    };

    this.utils = Object.assign({}, UTILS, { call });
  }

  addTemplate(name: string, template: string, options: { allowDuplicate?: boolean } = {}) {
    if (name in this.rawTemplates && !options.allowDuplicate) {
      throw new Error(`Template ${name} already defined`);
    }
    this.rawTemplates[name] = template;
  }

  getTemplate(name: string): Template {
    if (!(name in this.templates)) {
      const rawTemplate = this.rawTemplates[name];
      if (rawTemplate === undefined) {
        throw new Error(`Missing template: "${name}"`);
      }
      const templateFn = compileTemplate(rawTemplate, name);
      const template = templateFn(Blocks, this.utils);
      this.templates[name] = template;
    }
    return this.templates[name];
  }
}

export class App<T extends typeof Component = any> extends TemplateSet {
  Root: T;
  props: any;
  env: any = {};
  scheduler = new Scheduler(window.requestAnimationFrame.bind(window));
  root: OwlNode | null = null;
  dev: boolean = true;
  
  constructor(Root: T, props?: any) {
    super();
    this.Root = Root;
    this.props = props;
  }

  configure(params: any) {
    if (params.env) {
      this.env = params.env;
    }
  }

  mount(target: HTMLElement): Promise<InstanceType<T>> {
    if (!(target instanceof HTMLElement)) {
      throw new Error("Cannot mount component: the target is not a valid DOM element");
    }
    if (!document.body.contains(target)) {
      throw new Error("Cannot mount a component on a detached dom node");
    }
    const node = new OwlNode(this, this.Root, this.props);
    this.root = node;
    return node.mount(target);
  }

  destroy() {
    if (this.root) {
      this.root.destroy();
    }
  }
}
