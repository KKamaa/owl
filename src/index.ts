import { App } from "./app";
import { Component } from "./component";

export { App, Component };

export async function mount<T extends typeof Component>(
  C: T,
  target: HTMLElement
): Promise<InstanceType<T>> {
  const app = new App(C);
  return app.mount(target);
}

export { xml } from "./tags";
export { useState } from "./reactivity";
export { useRef } from "./refs";
export { EventBus } from "./event_bus";

export {
  onWillStart,
  onMounted,
  onBeforeUnmount,
  onWillUpdateProps,
  onBeforePatch,
  onPatched,
} from "./lifecycle_hooks";

export const __info__ = {};
