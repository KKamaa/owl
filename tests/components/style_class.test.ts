import { Component, mount, useState, xml } from "../../src";
import { fromName, makeTestFixture, nextTick, snapshotTemplateCode } from "../helpers";

let fixture: HTMLElement;

beforeEach(() => {
  fixture = makeTestFixture();
});

describe("style and class handling", () => {
  test("can set style and class on component", async () => {
    class Test extends Component {
      static template = xml`
          <div style="font-weight:bold;" class="some-class">world</div>
        `;
    }
    await mount(Test, { target: fixture });
    expect(fixture.innerHTML).toBe(`<div style="font-weight:bold;" class="some-class">world</div>`);
    snapshotTemplateCode(fromName(Test.template));
  });

  test("can set class on sub component", async () => {
    class Child extends Component {
      static template = xml`<div>child</div>`;
    }

    class Parent extends Component {
      static template = xml`<Child class="some-class" />`;
      static components = { Child };
    }
    snapshotTemplateCode(fromName(Parent.template));
    await mount(Parent, { target: fixture });
    expect(fixture.innerHTML).toBe(`<div class="some-class">child</div>`);
  });

  test("can set more than one class on sub component", async () => {
    class Child extends Component {
      static template = xml`<div>child</div>`;
    }

    class Parent extends Component {
      static template = xml`<Child class="a  b" />`;
      static components = { Child };
    }
    snapshotTemplateCode(fromName(Parent.template));
    await mount(Parent, { target: fixture });
    expect(fixture.innerHTML).toBe(`<div class="a b">child</div>`);
  });

  test("component class and parent class combine together", async () => {
    class Child extends Component {
      static template = xml`<div class="child">child</div>`;
    }

    class Parent extends Component {
      static template = xml`<Child class="from parent" />`;
      static components = { Child };
    }
    snapshotTemplateCode(fromName(Parent.template));
    await mount(Parent, { target: fixture });
    expect(fixture.innerHTML).toBe(`<div class="child from parent">child</div>`);
  });

  test("setting a class on a child component with text node", async () => {
    class Child extends Component {
      static template = xml`child`;
    }

    class Parent extends Component {
      static template = xml`<Child class="some-class" />`;
      static components = { Child };
    }
    snapshotTemplateCode(fromName(Parent.template));
    await mount(Parent, { target: fixture });
    expect(fixture.innerHTML).toBe(`child`);
  });

  test("can set class on sub sub component", async () => {
    class ChildChild extends Component {
      static template = xml`<div>childchild</div>`;
    }

    class Child extends Component {
      static template = xml`<ChildChild class="fromchild" />`;
      static components = { ChildChild };
    }

    class Parent extends Component {
      static template = xml`<Child class="fromparent" />`;
      static components = { Child };
    }
    snapshotTemplateCode(fromName(Parent.template));
    await mount(Parent, { target: fixture });
    expect(fixture.innerHTML).toBe(`<div class="fromchild fromparent">childchild</div>`);
  });

  test("can set class on multi root component", async () => {
    class Child extends Component {
      static template = xml`<div>a</div><span>b</span>`;
    }

    class Parent extends Component {
      static template = xml`<Child class="fromparent" />`;
      static components = { Child };
    }
    snapshotTemplateCode(fromName(Parent.template));
    await mount(Parent, { target: fixture });
    expect(fixture.innerHTML).toBe(`<div class="fromparent">a</div><span>b</span>`);
  });

  test("class on sub component, which is switched to another", async () => {
    class ChildA extends Component {
      static template = xml`<div>a</div>`;
    }
    class ChildB extends Component {
      static template = xml`<span>b</span>`;
    }
    class Child extends Component {
      static template = xml`<ChildA t-if="props.child==='a'"/><ChildB t-else=""/>`;
      static components = { ChildA, ChildB };
    }

    class Parent extends Component {
      static template = xml`<Child class="someclass" child="state.child" />`;
      static components = { Child };

      state = useState({ child: "a" });
    }
    snapshotTemplateCode(fromName(Parent.template));

    const parent = await mount(Parent, { target: fixture });
    expect(fixture.innerHTML).toBe(`<div class="someclass">a</div>`);
    parent.state.child = "b";
    await nextTick();
    expect(fixture.innerHTML).toBe(`<span class="someclass">b</span>`);
  });
});
