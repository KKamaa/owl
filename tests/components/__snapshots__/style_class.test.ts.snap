// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`style and class handling can set class on multi root component 1`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  class Block1 extends BStatic {
    static el = elem(\`<div>a</div>\`);
  }
  
  class Block2 extends BStatic {
    static el = elem(\`<span>b</span>\`);
  }
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = new BMulti(2);
    const b2 = b1.children[0] = new Block1();
    const b3 = b1.children[1] = new Block2();
    return b1;
  }
}"
`;

exports[`style and class handling can set class on multi root component 2`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = node.getChild(\`Child\`, {}, key + \`__1\`, ctx);
    b1.parentClass = toClassObj(\`fromparent\`);
    return b1;
  }
}"
`;

exports[`style and class handling can set class on sub component 1`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  class Block1 extends BStatic {
    static el = elem(\`<div>child</div>\`);
  }
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = new Block1();
    return b1;
  }
}"
`;

exports[`style and class handling can set class on sub component 2`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = node.getChild(\`Child\`, {}, key + \`__1\`, ctx);
    b1.parentClass = toClassObj(\`some-class\`);
    return b1;
  }
}"
`;

exports[`style and class handling can set class on sub sub component 1`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  class Block1 extends BStatic {
    static el = elem(\`<div>childchild</div>\`);
  }
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = new Block1();
    return b1;
  }
}"
`;

exports[`style and class handling can set class on sub sub component 2`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = node.getChild(\`ChildChild\`, {}, key + \`__1\`, ctx);
    b1.parentClass = toClassObj(\`fromchild\`);
    return b1;
  }
}"
`;

exports[`style and class handling can set class on sub sub component 3`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = node.getChild(\`Child\`, {}, key + \`__1\`, ctx);
    b1.parentClass = toClassObj(\`fromparent\`);
    return b1;
  }
}"
`;

exports[`style and class handling can set more than one class on sub component 1`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  class Block1 extends BStatic {
    static el = elem(\`<div>child</div>\`);
  }
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = new Block1();
    return b1;
  }
}"
`;

exports[`style and class handling can set more than one class on sub component 2`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = node.getChild(\`Child\`, {}, key + \`__1\`, ctx);
    b1.parentClass = toClassObj(\`a  b\`);
    return b1;
  }
}"
`;

exports[`style and class handling can set style and class on component 1`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  class Block1 extends BStatic {
    static el = elem(\`<div style=\\"font-weight:bold;\\" class=\\"some-class\\">world</div>\`);
  }
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = new Block1();
    return b1;
  }
}"
`;

exports[`style and class handling class is properly added on widget root el 1`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  class Block1 extends BStatic {
    static el = elem(\`<div class=\\"c\\"/>\`);
  }
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = new Block1();
    return b1;
  }
}"
`;

exports[`style and class handling class is properly added on widget root el 2`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  class Block1 extends BElem {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
    build() {
      this.anchors = new Array(1);
      this.anchors[0] = this.el.firstChild;
    }
  }
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = new Block1();
    const b2 = b1.children[0] = node.getChild(\`Child\`, {}, key + \`__1\`, ctx);
    b2.parentClass = toClassObj(\`a b\`);
    return b1;
  }
}"
`;

exports[`style and class handling class on sub component, which is switched to another 1`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  class Block1 extends BStatic {
    static el = elem(\`<div>a</div>\`);
  }
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = new Block1();
    return b1;
  }
}"
`;

exports[`style and class handling class on sub component, which is switched to another 2`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  class Block1 extends BStatic {
    static el = elem(\`<span>b</span>\`);
  }
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = new Block1();
    return b1;
  }
}"
`;

exports[`style and class handling class on sub component, which is switched to another 3`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = new BMulti(2);
    if (ctx['props'].child==='a') {
      b1.children[0] = node.getChild(\`ChildA\`, {}, key + \`__1\`, ctx);
    } else {
      b1.children[1] = node.getChild(\`ChildB\`, {}, key + \`__2\`, ctx);
    }
    return b1;
  }
}"
`;

exports[`style and class handling class on sub component, which is switched to another 4`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = node.getChild(\`Child\`, {child: ctx['state'].child}, key + \`__1\`, ctx);
    b1.parentClass = toClassObj(\`someclass\`);
    return b1;
  }
}"
`;

exports[`style and class handling component class and parent class combine together 1`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  class Block1 extends BStatic {
    static el = elem(\`<div class=\\"child\\">child</div>\`);
  }
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = new Block1();
    return b1;
  }
}"
`;

exports[`style and class handling component class and parent class combine together 2`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = node.getChild(\`Child\`, {}, key + \`__1\`, ctx);
    b1.parentClass = toClassObj(\`from parent\`);
    return b1;
  }
}"
`;

exports[`style and class handling empty class attribute is not added on widget root el 1`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  class Block1 extends BStatic {
    static el = elem(\`<span/>\`);
  }
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = new Block1();
    return b1;
  }
}"
`;

exports[`style and class handling empty class attribute is not added on widget root el 2`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  class Block1 extends BElem {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
    build() {
      this.anchors = new Array(1);
      this.anchors[0] = this.el.firstChild;
    }
  }
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = new Block1();
    const b2 = b1.children[0] = node.getChild(\`Child\`, {}, key + \`__1\`, ctx);
    return b1;
  }
}"
`;

exports[`style and class handling nothing happens if sub component is just some text 1`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = new BText(\`child\`);
    return b1;
  }
}"
`;

exports[`style and class handling nothing happens if sub component is just some text 2`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = node.getChild(\`Child\`, {}, key + \`__1\`, ctx);
    b1.parentClass = toClassObj(\`some-class\`);
    return b1;
  }
}"
`;

exports[`style and class handling setting a class on a child component with text node 1`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = new BText(\`child\`);
    return b1;
  }
}"
`;

exports[`style and class handling setting a class on a child component with text node 2`] = `
"function anonymous(Blocks, helpers
) {
  let {BElem, BText, BNode, BCollection, BHtml, BMulti, BStatic, BDispatch} = Blocks;
  let {elem, setText, withDefault, call, getTemplate, zero, callSlot, capture, toClassObj} = helpers;
  
  return function template(ctx, node, key = \\"\\") {
    const b1 = node.getChild(\`Child\`, {}, key + \`__1\`, ctx);
    b1.parentClass = toClassObj(\`some-class\`);
    return b1;
  }
}"
`;
